### PR TITLE
perf: avoid copying of creating memory dist calculator

### PR DESCRIPTION
**Pull request by BubbleCal**

### CODE REVIEW

File: rust/lance-index/src/vector/graph/memory.rs [hunk 1]
```diff
@@ -58,22 +58,25 @@ impl VectorStorage for InMemoryVectorStorage {
...
+            query: query.as_ptr(),
             metric_type: self.metric_type,
         })
     }
 }
 
 struct InMemoryDistanceCal {
     vectors: Arc<MatrixView<Float32Type>>,
-    query: Vec<f32>,
+    query: *const f32,
```
Comment (eddyxu) [inline rust/lance-index/src/vector/graph/memory.rs:69]: Can we use `&[f32]` with lifetime? Using raw pointer means that the lifetime of query is not guaranteed when the `distance()` method is called.

Comment (BubbleCal) [reply rust/lance-index/src/vector/graph/memory.rs:69]: i tried this but found that the `dist_calc` is created by the trait method of `VectorStorage`, so it's impossible to add lifetime for this.

any better way?

Comment (chebbyChefNEQ) [reply rust/lance-index/src/vector/graph/memory.rs:69]: if we don't care about locality
`Arc<[f32]>` works

Comment (wjones127) [reply rust/lance-index/src/vector/graph/memory.rs:69]: We should probably make queries some sort of `Arc<[T]>` early in the query, so it's cheaply clone-able.

Comment (chebbyChefNEQ) [reply rust/lance-index/src/vector/graph/memory.rs:69]: a bit more pedantic. What we want is exactly this: https://github.com/pytorch/pytorch/blob/main/c10/util/intrusive_ptr.h

but I'm not sure if there is an equivalent and easily implementable in rust

Comment (BubbleCal) [reply rust/lance-index/src/vector/graph/memory.rs]: the vector in `Query` is `Arc<dyn Array>`, changed the type from `&[f32]` to `ArrayRef`

File: rust/lance-linalg/src/matrix.rs [hunk 2]
```diff
@@ -147,6 +147,19 @@ impl<T: ArrowFloatType> MatrixView<T> {
...
+            "Centroid is not defined for transposed matrix."
+        );
+        if i >= self.num_rows() {
+            None
+        } else {
+            let dim = self.num_columns();
+            Some(self.data.slice(i * dim, dim))
+        }
+    }
+
```
Comment (wjones127) [inline rust/lance-linalg/src/matrix.rs:162]: Is this used?

**Review by wjones127 — APPROVED** [review]

File: rust/lance-linalg/src/matrix.rs (hunk 2 above)
Comment (BubbleCal) [reply rust/lance-linalg/src/matrix.rs:162]: yes, see
```suggestion
    pub fn vector(&self, id: u32) -> ArrayRef {
        self.vectors.row(id as usize).unwrap()
    }
```

**Pull request merged**